File: python/bellechasse/src/bellechasse/generators/fft_generator.py
from .generator import Generator


class FFTGenerator(Generator):

    def __init__(self, name, amp, subdivisions, memory_length, offset=0.5):
        super().__init__(name=name, amp=amp, period=0, phase=0, offset=offset)
        self.set_subdivisions_and_memory(subdivisions, memory_length)

        self.thres = 0

    def set_subdivisions_and_memory(self, subdivisions, memory_length):
        self.memory_length = memory_length
        self.subdivisions = subdivisions

        self.weighting = [0 for i in range(subdivisions)]
        self.stamps = [0 for i in range(memory_length)]
        self.memory = [[0 for i in range(subdivisions)] for j in range(memory_length)]

    def set_weighting(self, weighting):
        self.weighting = weighting.copy()

    def forward(self, values, millis):
        self.stamps[1:] = self.stamps[0:-1]
        self.stamps[0] = millis

        self.memory[1:] = self.memory_length[0:-1]

        for i in range(self.subdivisions):
            self.memory[0][i] = values[i] * self.weighting[i]

            if self.memory[0][i] < self.thres:
                self.memory[0][i] = 0
            else:
                self.memory[0][i] -= self.thres

    def value(self, millis):
        best_index = 0
        for i, _ in enumerate(self.stamps):
            best = abs(self.stamps[best_index] - millis)
            curr = abs(self.stamps[i] - millis)

            if curr < best:
                best_index = i

        fft_sum = 0
        for i in range(self.memory_length):
            fft_sum += self.memory[best_index][i]

        return fft_sum * self.amp + self.offset

File: processing/ParquetteLights/FFTGenerator.java
import processing.core.PApplet;

public class FFTGenerator extends Generator {

  int memoryLength;
  int subdivisions;
  float[] weighting;
  int[] stamps;
  float[][] memory;
  float thres;

  public FFTGenerator(PApplet p, String name, float amp, int subdivisions, int memoryLength) {
    super(p, name, amp, 0, 0);
    this.subdivisions = subdivisions;
    weighting = new float[subdivisions];
    setMemory(memoryLength);
    stamps = new int[memoryLength];
    setThreadshold(0);
  }

  public void forward(float[] values, int millis) {
    for (int i = 1; i < stamps.length; i++) {
    	stamps[i] = stamps[i-1];
    }

    stamps[0] = millis;

    for (int k = memoryLength-1; k >= 0; k--) {
	    if (k != 0) p.arraycopy(memory[k-1], memory[k]);
    }

    for (int i = 0; i < subdivisions; i++) {
      memory[0][i] = values[i]*weighting[i];

      if (memory[0][i] < thres) {
        memory[0][i] = 0;
      } else {
        memory[0][i] -= thres;
      }
    }
  }

  public float[] getWeighting() {
  	return weighting;
  }

  public void setWeights(float[] weighting) {
  	p.arraycopy(weighting, this.weighting);
  }

  public void setThreadshold(float thres) {
    this.thres = thres;
  }

  public void setMemory(int memoryLength) {
    this.memoryLength = memoryLength;
    memory = new float[memoryLength][subdivisions];
  }

  public float value(int millis) {
  	int bestIndex = -1;
  	for (int i = 0; i < stamps.length; i++) {
  		if (bestIndex == -1) {
  			bestIndex = i;
  			continue;
  		}

  		float best = p.abs(stamps[bestIndex] - millis);
  		float curr = p.abs(stamps[i] - millis);
  		if (curr < best) bestIndex = i;
  	}

  	float sum = 0.0f;
  	for (int i = 0; i < memory[0].length; i++) {
  		sum += memory[bestIndex][i];
  	}
  	return sum*amp;
  }
}
